crates/sui-node/src/lib.rs

```diff
@@ -248,6 +248,14 @@ impl SuiNode {
             &config.expensive_safety_check_config,
         );
 
+        // When we are opening the db table, the only time when it's safe to
+        // check SUI conservation is at genesis. Otherwise we may be in the middle of
+        // an epoch and the SUI conservation check will fail. This also initialize
+        // the expected_network_sui_amount table.
+        store
+            .expensive_check_sui_conservation()
```

Comment:
I don't understand this change. As in the comment, this is only safe to call at genesis (i.e. when the db table is empty). If it's not genesis, it could be in the middle of an epoch and it does not guarantee conservation at all.

Reply:
Is there a way to check for genesis at this point, or that the DB was otherwise opened? @dariorussi, the thing we missed when looking at the code yesterday is that within `AuthorityStore::open_inner` this call is guarded by a check that the database was previously empty, implying that the node is starting from scratch, and we will need some way to recreate that check at the new location:

https://github.com/MystenLabs/sui/blob/250221a6defc2fee77dc745b2f7b32a39693208d/crates/sui-core/src/authority/authority_store.rs#L244-L248

The reason we need to make this change is that the new implementation of the check will require access to the VM which is created as part of the `AuthorityPerEpochStore`, (which itself is created after the `AuthorityStore::open` that used to perform this check).

Reply:
let me look at this again then, I guess it has to be protected...

Reply:
@amnn I noticed :) and looking into how to workaround that.
@lxfind I think I can move the creation of the `AuthorityPerpetualTables` out of `AuthorityStore` and do the check there (before calling into `AuthorityStore`) and protect the check that way for both prod and testing.
I'll have code up soon, but the question @amnn asked is fair, if there was a simpler way

Reply:
@lxfind, Can we rely on the fact that the checks themselves use their own DB and initialise it if it's empty?

https://github.com/MystenLabs/sui/blob/250221a6defc2fee77dc745b2f7b32a39693208d/crates/sui-core/src/authority/authority_store.rs#L1599-L1619

Reply:
> @lxfind, Can we rely on the fact that the checks themselves use their own DB and initialise it if it's empty?

Not sure exactly what you meant.
But the earlier proposal of creating tables outside sounds reasonable thing to try

Reply:
Sorry, I meant "table" there instead of "DB".  I was referring to the fact that within the expensive check, there is a conditional (linked above) that checks whether there are already entries for `expected_storage_fund_imbalance` (and `expected_network_sui_amount`) in the perpetual tables.  If there are, then it does a conservation check, and if there are not, it initialises those tables.

>  If it's not genesis, it could be in the middle of an epoch and it does not guarantee conservation at all.

This bit still doesn't make sense to me:  How do we lose the conservation guarantee, assuming that even if the node is started mid-epoch, we initialise its conservation-related tables before it starts processing any transactions?

I'm assuming here that as long as we don't have any transactions in flight, we would expect the live object set at any time during an epoch to conserve SUI as well.

Reply:
The missing pieces are the gas rewards that's accumulated in Rust/db, not on-chain.
So if we do conservation check in the middle of an epoch, it's going to be very tricky to figure out how much gas reward that has accumulated so far.